extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -158,41 +170,46 @@ public void close() {
     }
 
     private void cancelOutstanding() {
-        for (int i = 0; i < requests.length; i++) {
-            if (requests[i] != null) {
-                requests[i].release();
-                requests[i] = null;
-            }
+        for (final Request request : localCache.values()) {
+            request.release();
         }
+        localCache.clear();
     }
 
     // --------------------------------------------------------------------------------------------------
 
     private Request getRequest(final long fragmentIndex) {
         final int cacheIdx = cacheIndex(fragmentIndex);
-        final Request request = requests[cacheIdx];
-        return request == null || !request.isFragment(fragmentIndex) ? null : request;
+        final Request locallyCached = localCache.get(cacheIdx);
+        if (locallyCached != null && locallyCached.isFragment(fragmentIndex)) {
+            return locallyCached;
+        }
+        final Request sharedCacheRequest = sharedCache.getRequest(uri, fragmentIndex);
```

Comment:
Based on the usage patterns, getRequest is only called _after_ getOrCreateRequest; in which case, I think that getRequest should _only_ be for the localCache, and not care about sharedCache.